cuBERT_topic_modelling/cuBERTopic.py

```diff
@@ -0,0 +1,563 @@
+        """Function to implement mean pooling on top of the AutoModel
+        See: https://www.sbert.net/examples/applications/computing-embeddings/README.html#sentence-embeddings-with-transformers
+
+        Args:
+            model_output \
+                (transformers.BaseModelOutputWithPoolingAndCrossAttentions): BERT model
+            attention_mask (torch.Tensor): torch.Tensor representing attention
+            mask values
+
+        Returns:
+            [torch.Tensor]: correct averaging of attention mask
+        """
+        # First element of model_output contains all token embeddings
+        token_embeddings = model_output[0]
+        input_mask_expanded = attention_mask.\
+            unsqueeze(-1).\
+            expand(token_embeddings.size()).\
+            float()
+        sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
+        sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
+        return sum_embeddings / sum_mask
+
+    def create_embeddings(self, sentences):
+        """Creates the sentence embeddings using SentenceTransformer
+
+        Args:
+            sentences (cudf.Series[str]): a cuDF Series of Input strings
+
+        Returns:
+            embeddings (cupy.ndarray): corresponding sentence
+            embeddings for the strings passed
+        """
+
+        cudf_tokenizer = SubwordTokenizer(
+            'voc_hash.txt',
+            do_lower_case=True
+        )
+
+        # Tokenize sentences
+        encoded_input_cudf = cudf_tokenizer(
+            sentences,
+            max_length=128,
+            max_num_rows=len(sentences),
+            padding='max_length',
+            return_tensors='pt',
+            truncation=True
+        )
+
+        # Load AutoModel from huggingface model repository
+        model_gpu = AutoModel.from_pretrained(
+            "sentence-transformers/all-MiniLM-L6-v2"
+        ).to(device)
+
+        # Delete the key and associated values we do not need
+        del encoded_input_cudf['metadata']
+
+        encoded_input_cudf['input_ids'] = self.fix_padding(
+            encoded_input_cudf['input_ids']
+        )
+        encoded_input_cudf['attention_mask'] = self.fix_padding(
+            encoded_input_cudf['attention_mask']
+        )
+
+        batch_size = 64
+
+        dataset = TensorDataset(
+            encoded_input_cudf['input_ids'],
+            encoded_input_cudf['attention_mask']
+        )
+
+        dataloader = DataLoader(
+            dataset,
+            batch_size=batch_size
+        )
+
+        model_o_ls = []
+        model_pooler_output = []
+        with torch.no_grad():
+            for data in dataloader:
+                mapping = {}
+                mapping['input_ids'] = data[0]
+                mapping['attention_mask'] = data[1]
+                model_o_ls.append(model_gpu(**mapping).last_hidden_state)
+                model_pooler_output.append(model_gpu(**mapping).pooler_output)
```

Comment:
Repeated model call.  We onlly need to call the model once. Model calls are extremely expensive.

Reply:
Good catch! Thanks!